### ISSUE TITLE
Can Adapter-Transformers be used for continued pretraining via adapters?

### ISSUE THREAD

**Issue by plamb-viso** [adapter-hub/adapter-transformers#526]
Can the `adapter-transformers` package be used for continued pretraining (vs just finetuning) via adapters as detailed in [Revisting Pretraining with Adapters](https://aclanthology.org/2021.repl4nlp-1.11.pdf)?

**Comment by skylersterling** [comment]
I would like an answer to this too.

**Comment by TimoImhof** [comment]
Hi, sorry for the delayed reply. I hope this is still relevant to you.
We are currently working on some big changes; therefore, I am getting back to you just now.

I took a look at this paper, so let me try to summarize their approach first so we are on the same page (and please correct me if I'm wrong somewhere):

First, let's talk about the steps of normal task-adaptive pretraining(TAPT):

- Step 1: A model is pretrained on some pretraining data with, e.g., masked language modelling (MLM)
- Step 2: The model is further pretrained, but now on the data of the target task. This is done to rule out distributional mismatches between pretraining and target task data.
- Step 3: The model is finetuned on the target task, e.g., text classification, so a classification layer is added on top of the pretrained model, and all model parameters, including the ones of the classification head, are finetuned on the task

Now the parameter-efficient variant of TAPT presented in this paper:

- Step 1: As before, a model is pretrained on some data
- Step 2: (Here is the change) Adapters are added to the model, to be precise language adapters (see [https://docs.adapterhub.ml/methods.html#language-adapters-invertible-adapters](url)  for an explanation). Now, these adapters are then pretrained on the target data (while the residual model weights are frozen).
- Step 3: After the second pretraining stage comes the finetuning to the target task. Let's take text classification again, so a classification head is added to the model and the weights of this head and adapters are finetuned on the target task (other model weights still frozen).

Let's create a small example to show how this can be done using adapter-transformers. We'll start with the second phase, the pretraining of adapters:

We load a pretrained adapter model from the Hub and add the language adapters and a masked language modelling head, which is our task for pretraining:

```
model = AutoAdapterModel.from_pretrained('roberta-base')
# Add a new adapter and masked language modeling head
model.add_adapter('parameter-eff-TAPT', config='pfeiffer+inv')
model.add_masked_lm_head('parameter-eff-TAPT')
model.set_active_adapters('parameter-eff-TAPT')
```
Next, you get some text data from a dataset, tokenize for MLM and divide it into training and validation splits:

```
# Load the SQuAD dataset
squad_dataset = load_dataset("squad")
tokenizer = RobertaTokenizer.from_pretrained('roberta-base')

# Define the tokenization function
def tokenize(batch):
    return tokenizer(batch['context'], padding=True, truncation=True)

eval_set = squad_dataset["validation"].map(tokenize, batched=True)
train_set = squad_dataset["train"].map(tokenize, batched=True)
```
For Training, we use the AdapterTrainer class. 
We specify our training arguments and create a data collator:
(For a complete guide to MLM see https://huggingface.co/docs/transformers/main/tasks/masked_language_modeling)

```
training_args = TrainingArguments( # here go your training arguments)
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=True, mlm_probability=0.15)
```
Last, we enable adapter training, which freezes the weights of the model except the adapter weights, create an instance of AdapterTrainer and start training:
```
model.train_adapter('parameter-eff-TAPT')
trainer = AdapterTrainer(
    model=model,
    args=training_args,
    train_dataset=train_set,
    eval_dataset=eval_set,
    tokenizer=tokenizer,
    data_collator=data_collator,
)
trainer.train()
```
After training, you can save the adapters as usual and again load them for finetuning on your target task. In the case of the example, we used SQuAD, so maybe for question answering.
Keep in mind that we also saved the MLM head together with the adapters, so for the finetuning of the target task it must be deleted and a new head must be added.
```
# Third stage
model = AutoAdapterModel.from_pretrained('roberta-base')
# Load the pretrained adapter and model head
model.load_adapter("/your/path/to/adapter")
# With the print statements you can see if a head is also added to the model
print(model.heads)
# This is the command for deleting the head
model.delete_head('parameter-eff-TAPT')
print(model.heads)
# Now add your head necessary for your target task
model.add_qa_head('parameter-eff-TAPT')
print(model.heads)
# Start the finetuning of your adapters for the target task
model.set_active_adapters('parameter-eff-TAPT')
# Here comes code for the finetuning
```
I hope this helps you; if you have further questions, don't hesitate to ask! :)
